systemtest/src/test/java/io/strimzi/systemtest/migration/MigrationST.java

```diff
@@ -0,0 +1,265 @@
+
+        // the controller pods will not be up and running, because we are using the ZK nodes as controllers, they will be created once the migration starts
+        // creating it here (before KafkaTopics) to correctly delete KafkaTopics and prevent stuck because UTO cannot connect to controllers
+        resourceManager.createResourceWithoutWait(extensionContext,
+            KafkaNodePoolTemplates.kafkaNodePoolWithControllerRoleAndPersistentStorage(testStorage.getNamespaceName(), testStorage.getControllerPoolName(), testStorage.getClusterName(), 3).build());
+
+        // at this moment, everything should be ready, so we should ideally create some topics and send + receive the messages (to have some data present in Kafka + metadata about topics in ZK)
+        LOGGER.info("Creating two topics for immediate and continuous message transmission and KafkaUser for the TLS");
+        resourceManager.createResourceWithWait(extensionContext,
+            KafkaTopicTemplates.topic(testStorage).build(),
+            KafkaTopicTemplates.topic(testStorage.getClusterName(), continuousTopicName, testStorage.getNamespaceName()).build(),
+            KafkaUserTemplates.tlsUser(testStorage).build()
+        );
+
+        // sanity check that kafkaMetadataState shows ZooKeeper
+        String currentKafkaMetadataState = KafkaResource.kafkaClient().inNamespace(testStorage.getNamespaceName()).withName(testStorage.getClusterName()).get().getStatus().getKafkaMetadataState();
+        assertThat(currentKafkaMetadataState, is(KafkaMetadataState.ZooKeeper.name()));
+
+        // start continuous clients and do the immediate message transmission
+        resourceManager.createResourceWithWait(extensionContext,
+            continuousClients.producerStrimzi(),
+            continuousClients.consumerStrimzi(),
+            immediateClients.producerTlsStrimzi(testStorage.getClusterName()),
+            immediateClients.consumerTlsStrimzi(testStorage.getClusterName())
+        );
+
+        ClientUtils.waitForClientsSuccess(testStorage.getProducerName(), testStorage.getConsumerName(), testStorage.getNamespaceName(), testStorage.getMessageCount());
+
+        // starting the migration
+        LOGGER.info("Starting the migration process");
+
+        LOGGER.info("1. applying the {} annotation with value: {}", Annotations.ANNO_STRIMZI_IO_KRAFT, "migration");
+        KafkaResource.replaceKafkaResourceInSpecificNamespace(testStorage.getClusterName(),
+            kafka -> kafka.getMetadata().getAnnotations().put(Annotations.ANNO_STRIMZI_IO_KRAFT, "migration"), testStorage.getNamespaceName());
+
+        LOGGER.info("2. waiting for controller Pods to be up and running");
+        PodUtils.waitForPodsReady(testStorage.getNamespaceName(), controllerSelector, 3, true);
+        Map<String, String> controllerPodsSnapshot = PodUtils.podSnapshot(testStorage.getNamespaceName(), controllerSelector);
+
+        LOGGER.info("3. waiting for first rolling update of broker Pods");
```

Comment:
Getting info from log would cause race conditions FMPOV (the log doesn't contain info in time, it doesn't match the pattern, or whatever). So having a comment here why it's rolling would be better.

Reply:
Cool